HB/structure.elpi

```diff
@@ -310,8 +310,41 @@ declare-coercion SortProjection ClassProjection
   log.coq.env.add-const-noimplicits SName SCoeBody STy @transparent! SC,
   log.coq.coercion.declare (coercion (const SC) 0 StructureF (grefclass StructureT)),
   log.coq.CS.declare-instance (const SC), % TODO: API in Elpi, take a @constant instead of gref
+
+  if-verbose (coq.say "HB: declare coercion path compression rules"),
+
+  findall-classes All,
+  CurrentTgtClass = (class TC StructureT TMLwP),
+  std.filter All (sub-class? CurrentTgtClass) AllTgtSuper,
+  std.map AllTgtSuper class_structure AllTgtSuperStructures,
+
+  mk-compression-clauses StructureF StructureT AllTgtSuperStructures AllCompressionClauses,
+  std.forall AllCompressionClauses (c\ log.coq.env.accumulate current "hb.db" (clause _ _ c)),
 ].
 
+
+pred mk-compression-clauses i:gref, i:gref, i:list gref, o:list prop.
+mk-compression-clauses StructureF StructureT [] [].
+mk-compression-clauses StructureF StructureT [StructureE|Rest] [Clause|Clauses]:- std.do! [
+  coq.coercion.db-for (grefclass StructureF) (grefclass StructureT) [pr C1 Nparams1],
+  coq.coercion.db-for (grefclass StructureT) (grefclass StructureE) [pr C2 Nparams2],
+  coq.coercion.db-for (grefclass StructureF) (grefclass StructureE) [pr C3 Nparams3],
+  coq.mk-app (global C1) {coq.mk-n-holes Nparams1} F,
+  coq.mk-app (global C2) {coq.mk-n-holes Nparams2} G,
+  coq.mk-app (global C3) {coq.mk-n-holes Nparams3} H,
+  RuleSkel = {{ fun x => lp:G (lp:F x) = lp:H x}},
+  std.assert-ok! (coq.elaborate-skeleton RuleSkel _ Rule) "coercion composition fails",
+  (pi X L\ fold-map X L X [X|L] :- var X, not(std.exists L (same_var X))) => fold-map Rule [] Rule Holes,
+  mk-compression-clause Holes Rule Clause,
+  mk-compression-clauses StructureF StructureT Rest Clauses,
+].
+
+pred mk-compression-clause i:list term, i:term, o:prop.
+mk-compression-clause [] (fun _ _ x\ app[_,_,LHS x,RHS x]) (pi x\ C x) :-
+  pi x\ copy (LHS x) (L x), copy (RHS x) (R x), C x = (pi tmp\ compress (L x) tmp :- !, copy (R x) tmp, !).
+mk-compression-clause [UV|Rest] T (pi v\ R v) :-
+  pi v\ (pi U\ copy U v :- same_var U UV, !) => mk-compression-clause Rest T (R v).
```

Comment:
@pi8027 these lines are the cheapest way I could find to make a clause out of a term quantifying unresolved holes

Reply:
It goes from the result of elaborating `{{ fun x => lp:G (lp:F x) = lp:H x}}}` to `pi holes... x tmp\ compress (elab_G_F x) tmp :- copy (elab_H x) tmp` where elab_G_F and elab_H may contain holes